src/snapping/index.js

```diff
@@ -0,0 +1,271 @@
+const turf = require("@turf/turf");
+const throttle = require("lodash.throttle");
+const {
+  getBufferLayerId,
+  getBufferLayer,
+  vertexIfClose,
+  featureWrapperOnPoint,
+  selectedFeatureIsPoint,
+  notPointFilter,
+  notSelectedFeatureFilter
+} = require("./util");
+
+class Snapping {
+  constructor(ctx) {
+    this.map = ctx.map;
+    this.snappedFeature = null;
+    this.snappedGeometry = null;
+    this.bufferLayers = [];
+    this.snapLayers = ctx.options.snapLayers;
+    this.snapFeatureFilter = ctx.options.snapFeatureFilter;
+    this.snapDistance = ctx.options.snapDistance;
+    this.store = ctx.store;
+    // this is the amount the endpoints are preferenced as snap points. and is related to the angle between the hover point, the nearest point and the endpoint
+    this.vertexPullFactor = Math.sqrt(2);
+
+    this._mouseoverHandler = this._mouseoverHandler.bind(this);
+    this._mouseoutHandler = this._mouseoutHandler.bind(this);
+    this.refreshSnapLayers = this.refreshSnapLayers.bind(this);
+    this.setSnapLayers = this.setSnapLayers.bind(this);
+    this.setSnapFeatureFilter = this.setSnapFeatureFilter.bind(this);
+    this.clearSnapCoord = this.clearSnapCoord.bind(this);
+
+    this.initialize();
+    this.attachApi(ctx);
+  }
+
+  initialize() {
+    this.map.on("styledata", () => {
+      this._updateSnapLayers();
+    });
+    this.map.on("draw.update", () => {
+      this.clearSnapCoord();
+    });
+  }
+
+  /** OUTWARD FACING METHODS */
+
+  attachApi(ctx) {
+    // To whom so ever has beef with this, I'm with you, but without re-designing things on a greater scale... this is how it is.
+    ctx.api.refreshSnapLayers = this.refreshSnapLayers;
+    ctx.api.setSnapLayers = this.setSnapLayers;
+    ctx.api.setSnapFeatureFilter = this.setSnapFeatureFilter;
+    ctx.api.clearSnapCoord = this.clearSnapCoord;
+  }
+
+  refreshSnapLayers() {
+    this._updateSnapLayers();
+  }
+
+  setSnapLayers(snapLayers) {
+    this.snapLayers = snapLayers;
+    this._updateSnapLayers();
+  }
+  setSnapFeatureFilter(snapFeatureFilter) {
+    this.snapFeatureFilter = snapFeatureFilter;
+  }
+
+  clearSnapCoord() {
+    const source = this.map.getSource("_snap_vertex");
+    if (source) {
+      source.setData({ type: "FeatureCollection", features: [] });
+    }
+  }
+
+  enableSnapping() {
+    console.log("snap on");
```

Comment:
the console log is a hold over from the old code. These sort of "mode on" console logs seem somewhat standard with the repo, so I left it